### PR TITLE
Browse namespace tree

### PR DESCRIPTION
**Pull request by hojberg**
## Overview
Fetch a namespace when expanding in the tree view

Feb-08-2021 10-33-41

## Implementation notes
A few new/reshaped modules came out of this effort:

* `NamespaceListing`: The main data structure, somewhat reflecting the API, that drives the tree view
* `FullyQualifiedName`: barely used right now, but expected to grow and definition rendering is introduced
* `Hash`: general Unison hash specifics
* `HashSet`: a `Set` wrapper, since Elm doesn't allow custom types to be comparable (needed for things like `Set.member` and `Set.toggle`). `HashSet` is used for tracking the state that controls which `NamespaceListing`s are expanded.
* A UI module for common UI views - eventually will include design system things like buttons, input fields etc to make sure a consistent styling and behavior is applied universally.

## Interesting/controversial decisions
I went back and forth a bunch (as perhaps evident from the commits) on the right data structure for the `NamespaceListings`. A tree is a natural choice, but the async nature of it (each sub tree is loaded when the user clicks to expand as opposed to ahead of time) made it a bit more interesting.

I contemplated having a more flat structure - Like a `Dict` of `RemoteData NamespaceListing` and each listing merely having a `Hash` reference to its children. This would make updating the state for a loaded listing simpler, but the tree traversal performance is likely fine for now.

## Loose ends
This is currently using the `namespaceName` as a `Hash` for Subnamespaces, which means expansion below the first level will fail as it requests the server with an incomplete identifier (it works for the top level because "base" is queryable, but "List" with out "base.List" or a `Hash` is not).

### CODE REVIEW

File: src/App.elm [hunk 1]
```diff
@@ -71,133 +76,176 @@ update msg model =
...
+                            HashSet.toggle hash
+                                model.expandedNamespaceListings
+                    }
+
+                cmd =
+                    if shouldExpand then
+                        fetchSubNamespaceListing hash
+
+                    else
+                        Cmd.none
```
Comment (hojberg) [inline src/App.elm:96]: We don't want to fetch on collapse

File: src/App.elm [hunk 2]
```diff
@@ -71,133 +76,176 @@ update msg model =
...
-        slash =
-            span [ class "slash" ] [ text "/" ]
-    in
-    label
-        [ class "definition-namespace-path" ]
-        (List.intersperse slash namespaceLinks)
+        listings =
+            case namespaceRoot of
+                Success (NamespaceListing _ fqn content) ->
+                    viewNamespaceListingContent expandedNamespaceListings content
```
Comment (hojberg) [inline src/App.elm:225]: The root should **not** have a node on its own so we render the children directly instead

File: src/Main.elm [hunk 3]
```diff
@@ -10,5 +10,5 @@ main =
         { init = App.init
         , update = App.update
         , view = App.view
-        , subscriptions = App.subscriptions
+        , subscriptions = \_ -> Sub.none
```
Comment (hojberg) [inline src/Main.elm:13]: We don't use Subscriptions right now.

File: src/NamespaceListing.elm [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+type DecodedNamespaceChild
+    = SubNamespace NamespaceListing
+    | SubDefinition DefinitionListing
+
+
+decodeSubNamespace : Json.Decoder DecodedNamespaceChild
+decodeSubNamespace =
+    Json.map3 NamespaceListing
+        -- TODO namespaceName should be namespaceHash
+        (field "namespaceName" Json.string |> andThen (Hash >> Json.succeed))
```
Comment (hojberg) [inline src/NamespaceListing.elm:70]: @runarorama would it be possible to had both an fqn and a hash to any NamespaceListingChild?

**Comment by pchiusano** [comment]
Cool. FYI, this package may be of interest, lets you create maps and sets with custom hashing and equality functions: https://package.elm-lang.org/packages/nikita-volkov/hashing-containers/latest/ Based on [HAMTs](https://en.wikipedia.org/wiki/Hash_array_mapped_trie) which are likely to be faster than Elm's default `Set` type, which is comparison based. Especially for long-ish keys like 512 bit hashes.

**Comment by hojberg** [comment]
@pchiusano I think I might explore that library in a separate effort if thats ok.

**Comment by pchiusano** [comment]
Yup no prob, was just putting it on your radar.

**Pull request merged**